### PR TITLE
Allow smoother resets

### PR DESCRIPTION
**Pull request by kobytg17**
1. Currently after a reset (or init), since p_error_last is_ set to 0, the first computeCommand is calculating
a false large derivative error. Proposed fix: only calculate error_dot if p_error_last_ contains a real value.

2. Allow setting initial i_error. This allows PID to be used in a smoother manner in systems switching from
manual to automatic modes. [Reference](http://brettbeauregard.com/blog/2011/04/improving-the-beginner%E2%80%99s-pid-initialization/)

3. Improve cases where dt==0 or invalid errors were given (return last cmd_ instead of 0)

### CODE REVIEW

File: src/pid.cpp [hunk 1]
```diff
@@ -299,15 +306,16 @@ double Pid::computeCommand(double error, ros::Duration dt)
 {
 
   if (dt == ros::Duration(0.0) || std::isnan(error) || std::isinf(error))
-    return 0.0;
+    return cmd_;
```
Comment (bmagyar) [inline src/pid.cpp]: What does this improve? It is unclear what a PID command should be in any of the cases in the `if` statement. The only reason we aren't throwing an exception here is to keep a simplistic, controller-friendly interface. We could change the return to `-inf` or `nan` if you wish ;)

**Comment by bmagyar** [comment]
ping @ipa-mdl

File: src/pid.cpp [hunk 2]
```diff
@@ -299,15 +306,16 @@ double Pid::computeCommand(double error, ros::Duration dt)
...
 
   double error_dot = d_error_;
 
   // Calculate the derivative error
-  if (dt.toSec() > 0.0)
+  if (dt.toSec() > 0.0 && valid_p_error_last_)
   {
     error_dot = (error - p_error_last_) / dt.toSec();
     p_error_last_ = error;
+    valid_p_error_last_ = true;
```
Comment (ipa-mdl) [inline src/pid.cpp]: I don't understand this change. This code will only be reached if `valid_p_error_last_` is already `true`.
If it is `false`, only `reset()` can set it to `true` again.

Comment (kobytg17) [reply src/pid.cpp]: My bad, I don't know how this slipped in this way.. Sorry
Added a test

**Comment by kobytg17** [comment]
Can we merge this please?

**Comment by kobytg17** [comment]
Do we want to merge, or should I close this PR? @ipa-mdl @bmagyar

**Comment by bmagyar** [comment]
Merging for Melodic and Noetic

**Pull request merged**